File: py/openai/src/agentloop_openai/_wrap.py
"""
wrap_openai — drop-in wrapper for an OpenAI client.

Returns a wrapped client where `.chat.completions.create()` fires
AgentLoop hooks automatically. Every other namespace (embeddings,
files, audio, etc.) passes through to the original client unchanged.

The original client is NOT mutated — wrap() returns a distinct object.
This matches the Langfuse/Helicone convention and avoids surprises
when both wrapped and unwrapped references exist in the same codebase.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from agentloop import AgentLoop

from ._ask import PerCallOptions, WrapOptions, ask_with_agentloop

if TYPE_CHECKING:
    from openai import OpenAI


class _WrappedCompletions:
    """Proxy for `client.chat.completions` with a replaced `create` method."""

    def __init__(
        self, real_completions: Any, config: WrapOptions
    ) -> None:
        self._real = real_completions
        self._config = config
        # Pull the underlying client reference from the completions object.
        # OpenAI's SDK structure means completions has an _client attribute
        # but we don't rely on that — ask_with_agentloop takes the client
        # explicitly, so we pass it through the config or a closure.
        # Here we stash it for the wrapped create() to use.
        self._client = getattr(real_completions, "_client", None)

    def create(
        self, *, agentloop: PerCallOptions | dict[str, Any] | None = None, **kwargs: Any
    ) -> Any:
        """Intercepted create. Strips `agentloop` kwarg before forwarding."""
        per_call = self._coerce_per_call(agentloop)

        # If the caller passed skip or there's no client reference, just
        # forward without any hooks.
        if per_call.skip:
            return self._real.create(**kwargs)

        # Need the outer client to re-run create() with augmented messages.
        # We work around the missing _client by delegating back to _real
        # with the same-shaped call. ask_with_agentloop uses
        # client.chat.completions.create — we construct a minimal shim.
        messages = kwargs.pop("messages", [])
        return ask_with_agentloop(
            _ClientShim(self._real),  # type: ignore[arg-type]
            messages=messages,
            per_call=per_call,
            config=self._config,
            **kwargs,
        )

    @staticmethod
    def _coerce_per_call(
        agentloop: PerCallOptions | dict[str, Any] | None,
    ) -> PerCallOptions:
        if agentloop is None:
            return PerCallOptions()
        if isinstance(agentloop, PerCallOptions):
            return agentloop
        if isinstance(agentloop, dict):
            return PerCallOptions(**agentloop)
        raise TypeError(
            f"agentloop must be PerCallOptions or dict, got {type(agentloop).__name__}"
        )

    def __getattr__(self, name: str) -> Any:
        # Delegate everything else (create_with_completion, etc.) to the
        # real completions object.
        return getattr(self._real, name)


class _ClientShim:
    """Minimal object with .chat.completions.create — just enough for
    ask_with_agentloop to call back into the underlying real completions.

    We do this instead of trying to reconstruct the whole OpenAI client
    object, which has many fields we don't need to touch.
    """

    def __init__(self, real_completions: Any) -> None:
        self.chat = _ClientShimChat(real_completions)


class _ClientShimChat:
    def __init__(self, real_completions: Any) -> None:
        self.completions = real_completions


class _WrappedChat:
    """Proxy for `client.chat`. Only `.completions` is overridden."""

    def __init__(self, real_chat: Any, config: WrapOptions) -> None:
        self._real = real_chat
        self.completions = _WrappedCompletions(real_chat.completions, config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._real, name)


class WrappedOpenAI:
    """Proxy for the full OpenAI client. Delegates everything except
    `chat` through `__getattr__` so embeddings, files, images, etc.
    all pass through unchanged.
    """

    def __init__(self, client: OpenAI, config: WrapOptions) -> None:
        self._real = client
        self.chat = _WrappedChat(client.chat, config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._real, name)


def wrap_openai(
    client: OpenAI,
    *,
    loop: AgentLoop,
    inject_memories: Any = None,
    detect_signals: Any = None,
    search_limit: int = 3,
    search_tags: list[str] | None = None,
    only_log_when_signaled: bool = False,
) -> OpenAI:
    """Wrap an OpenAI client so `chat.completions.create` calls fire
    AgentLoop hooks automatically.

    The return value is typed as OpenAI for drop-in compatibility. The
    only observable difference: create() accepts an extra `agentloop`
    kwarg for per-call overrides.

    Example:

        from openai import OpenAI
        from agentloop import AgentLoop
        from agentloop_openai import wrap_openai

        openai = wrap_openai(
            OpenAI(),
            loop=AgentLoop(api_key="ak_..."),
        )

        resp = openai.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "hi"}],
            agentloop={"user_id": "u_123"},  # per-call override
        )
    """
    if not loop:
        raise ValueError("wrap_openai: loop is required")

    config = WrapOptions(
        loop=loop,
        inject_memories=inject_memories,
        detect_signals=detect_signals,
        search_limit=search_limit,
        search_tags=search_tags or [],
        only_log_when_signaled=only_log_when_signaled,
    )
    return WrappedOpenAI(client, config)  # type: ignore[return-value]

File: js/openai/src/wrap.ts
/**
 * wrapOpenAI — the drop-in proxy wrapper.
 *
 * Returns a Proxy around the OpenAI client that intercepts
 * `chat.completions.create` calls and routes them through AgentLoop.
 * Every other path (openai.embeddings, openai.files, etc.) passes
 * through unchanged.
 *
 * Per-call options are passed via a dedicated `agentloop` field on the
 * create() params object. The wrapper strips that field before
 * forwarding to OpenAI, which rejects unknown top-level fields.
 */

import type OpenAI from "openai";
import type {
  ChatCompletion,
  ChatCompletionChunk,
  ChatCompletionCreateParams,
  ChatCompletionCreateParamsNonStreaming,
  ChatCompletionCreateParamsStreaming,
} from "openai/resources/chat/completions";
import type { Stream } from "openai/streaming";

import { askWithAgentLoop } from "./ask.js";
import { askWithAgentLoopStream } from "./stream.js";
import type { AgentLoopOpenAIOptions, PerCallOptions } from "./ask.js";

/**
 * Augmented params accepted by the wrapped client. Same as OpenAI's
 * normal params plus an `agentloop` field for per-call overrides.
 */
export type WrappedCreateParams =
  | (ChatCompletionCreateParamsNonStreaming & { agentloop?: PerCallOptions })
  | (ChatCompletionCreateParamsStreaming & { agentloop?: PerCallOptions });

/**
 * Wrap an OpenAI client so chat.completions.create calls fire
 * AgentLoop pre/post hooks automatically.
 *
 * The returned object is typed as OpenAI for drop-in compatibility.
 * The only observable difference: create() accepts an extra `agentloop`
 * field for per-call overrides.
 *
 * @example
 *   const openai = wrapOpenAI(new OpenAI(), { loop });
 *   const resp = await openai.chat.completions.create({
 *     model: "gpt-4o-mini",
 *     messages: [{ role: "user", content: "hello" }],
 *     agentloop: { userId: "u_123" },
 *   });
 */
export function wrapOpenAI(
  client: OpenAI,
  config: AgentLoopOpenAIOptions
): OpenAI {
  if (!config || !config.loop) {
    throw new Error("wrapOpenAI: config.loop is required");
  }

  const wrappedCreate = ((params: WrappedCreateParams, _options?: unknown) => {
    // Separate our field from OpenAI's params before forwarding.
    const { agentloop, ...openaiParams } = params as WrappedCreateParams & {
      agentloop?: PerCallOptions;
    };

    if ("stream" in openaiParams && openaiParams.stream === true) {
      return askWithAgentLoopStream(
        client,
        openaiParams as ChatCompletionCreateParamsStreaming,
        agentloop,
        config
      );
    }

    return askWithAgentLoop(
      client,
      openaiParams as ChatCompletionCreateParamsNonStreaming,
      agentloop,
      config
    );
  }) as typeof client.chat.completions.create;

  // Build a new object that looks like the original client but with our
  // create() swapped in. We intentionally do NOT mutate the original
  // client — callers who want to keep both a wrapped and unwrapped
  // reference can, without the Proxy leaking into other code paths.
  //
  // This is the Langfuse/Helicone convention: wrap returns a distinct
  // object; the original stays pristine.
  //
  // We use a Proxy for the top level so every other property (embeddings,
  // files, audio, etc.) passes through to the original client
  // unchanged, with correct `this` binding.
  return new Proxy(client, {
    get(target, prop, receiver) {
      if (prop === "chat") {
        // Return a synthetic chat namespace with a swapped-in completions.
        return {
          // Preserve all other methods on `chat` (currently none, but
          // future-proof against OpenAI adding e.g. chat.something).
          ...target.chat,
          completions: {
            ...target.chat.completions,
            create: wrappedCreate,
          },
        };
      }
      return Reflect.get(target, prop, receiver);
    },
  });
}
